### ISSUE TITLE
[data grid] Quick filter performance

### ISSUE THREAD

**Issue by romgrk** [mui/mui-x#9167]
Here is a summary of the issues for quick filters performance, to be completed after #9120 is merged.

### 1. Date operations

Our date formatters are absurdly slow. They dominate any profile where they are used for quick filtering. A partial fix would be to create a reusable `Intl.DateTime` instance instead of calling `.toLocaleString()` (which creates a new `Intl.DateTime` under the hood).

https://github.com/mui/mui-x/blob/bd291b3ab9032c8259bdbf78ff1be40b616fad74/packages/grid/x-data-grid/src/colDef/gridDateColDef.ts#L35

Additionally, replacing the native formatter by pure javascript also improves performance a lot. I don't think we can do this ourselves, but we can add a note in the docs for users who are interested in getting more performance.

E.g. 
```tsx
{
  valueFormatter: ({ value }) => `${value.getFullYear()}-${value.getMonth() + 1}-...`
}
```

### 2. Ignore non-filterable columns

We have a `.filterable` flag on coldefs but we don't use it for quick filters. Currently the only way to exclude a column from quick filtering is to unset `.getApplyQuickFilterFn`.

### Other

The other improvements to make are to apply the same changes as in #9120, in particular introducing new V7 API that are less resource-hungry (for formatters in particular), and to use `eval()` or `new Function()` to monomorphize our functions & eliminate dynamic key loads & stores.

**Comment by michelengelen** [comment]
Closing this in favor of #9657, which tracks the quick filter performance work with the discussion, the benchmarks, and the current plan (lower debounce, chunked and interruptible filtering, INP). Two issues for one workstream split the signal, so let us keep one.

Carrying the open items from this checklist over to #9657, since none of them shipped yet:

- [ ] Date formatters: reuse a single `Intl.DateTimeFormat` instance instead of calling `value.toLocaleString()` per cell ([still current](https://github.com/mui/mui-x/blob/master/packages/x-data-grid/src/colDef/gridDateColDef.ts)), and document the pure-JS formatter escape hatch for perf-sensitive users.
- [ ] Quick filtering ignores `colDef.filterable`: the column selection only checks visibility and `getApplyQuickFilterFn`, so `filterable: false` columns still run through the quick filter.

The third item (monomorphization via `new Function()`) predates the v7 API work and stays an optional exploration. What shipped from this era: #9712 lowered the debounce to 150ms and made it configurable through `filterDebounceMs`.

Please follow #9657 for this workstream — I posted the two items above there so they stay actionable.
